File: Python/findNumOfValidWords.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: findNumOfValidWords.py
@Function: 猜字谜
@Link: https://leetcode-cn.com/problems/number-of-valid-words-for-each-puzzle/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-30
"""


from typing import List
import collections


class Solution:
    def findNumOfValidWords(self, words: List[str], puzzles: List[str]) -> List[int]:
        frequency = collections.Counter()

        for word in words:
            mask = 0
            for ch in word:
                mask |= (1 << (ord(ch) - ord("a")))
            if str(bin(mask)).count("1") <= 7:
                frequency[mask] += 1

        ans = list()
        for puzzle in puzzles:
            total = 0

            # 枚举子集方法一
            # for choose in range(1 << 6):
            #     mask = 0
            #     for i in range(6):
            #         if choose & (1 << i):
            #             mask |= (1 << (ord(puzzle[i + 1]) - ord("a")))
            #     mask |= (1 << (ord(puzzle[0]) - ord("a")))
            #     if mask in frequency:
            #         total += frequency[mask]

            # 枚举子集方法二
            mask = 0
            for i in range(1, 7):
                mask |= (1 << (ord(puzzle[i]) - ord("a")))

            subset = mask
            while subset:
                s = subset | (1 << (ord(puzzle[0]) - ord("a")))
                if s in frequency:
                    total += frequency[s]
                subset = (subset - 1) & mask

            # 在枚举子集的过程中，要么会漏掉全集 mask，要么会漏掉空集
            # 这里会漏掉空集，因此需要额外判断空集
            if (1 << (ord(puzzle[0]) - ord("a"))) in frequency:
                total += frequency[1 << (ord(puzzle[0]) - ord("a"))]

            ans.append(total)

        return ans


# -----------------------------
if __name__ == "__main__":
    words = ["aaaa", "asas", "able", "ability", "actt", "actor", "access"]
    puzzles = ["aboveyz", "abrodyz", "abslute",
               "absoryz", "actresz", "gaswxyz"]

    solution = Solution()
    find_number_valid_words = solution.findNumOfValidWords(words, puzzles)
    print(f"The solutio of this problem is {find_number_valid_words}")

File: C++/findNumOfValidWords.cpp
/**
 * @File    : findNumOfValidWords.cpp
 * @Brief   : 猜字谜
 * @Link    : https://leetcode-cn.com/problems/number-of-valid-words-for-each-puzzle/
 * @Author  : Wei Li
 * @Date    : 2021-08-30
*/

#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>

class Solution
{
public:
    std::vector<int> findNumOfValidWords(std::vector<std::string> &words, std::vector<std::string> &puzzles)
    {
        std::unordered_map<int, int> frequency;

        for (const std::string &word : words)
        {
            int mask = 0;
            for (char ch : word)
            {
                mask |= (1 << (ch - 'a'));
            }
            if (__builtin_popcount(mask) <= 7)
            {
                ++frequency[mask];
            }
        }

        std::vector<int> ans;
        for (const std::string &puzzle : puzzles)
        {
            int total = 0;

            // 枚举子集方法一
            // for (int choose = 0; choose < (1 << 6); ++choose) {
            //     int mask = 0;
            //     for (int i = 0; i < 6; ++i) {
            //         if (choose & (1 << i)) {
            //             mask |= (1 << (puzzle[i + 1] - 'a'));
            //         }
            //     }
            //     mask |= (1 << (puzzle[0] - 'a'));
            //     if (frequency.count(mask)) {
            //         total += frequency[mask];
            //     }
            // }

            // 枚举子集方法二
            int mask = 0;
            for (int i = 1; i < 7; ++i)
            {
                mask |= (1 << (puzzle[i] - 'a'));
            }
            int subset = mask;
            do
            {
                int s = subset | (1 << (puzzle[0] - 'a'));
                if (frequency.count(s))
                {
                    total += frequency[s];
                }
                subset = (subset - 1) & mask;
            } while (subset != mask);

            ans.push_back(total);
        }
        return ans;
    }
};

template <typename T>
void DispalyElement(const T &container)
{
    for (auto element : container)
    {
        std::cout << element << " ";
    }
    std::cout << std::endl;
}

// ------------------------------
int main(int argc, char **argv)
{
    std::vector<std::string> words{"aaaa", "asas", "able", "ability", "actt", "actor", "access"};
    std::vector<std::string> puzzles{"aboveyz", "abrodyz", "abslute", "absoryz", "actresz", "gaswxyz"};

    Solution solution;
    auto find_number_valid_words = solution.findNumOfValidWords(words, puzzles);
    DispalyElement(find_number_valid_words);

    return 0;
}
